content/chapitres/ci.adoc

```diff
@@ -164,29 +219,57 @@ include::../code-samples/gh-actions/say-hello-full.yml[tags="common,checkout,sho
+** link:https://www.destroyallsoftware.com/talks/wat[WAT 🤪, window="_blank"] ?
 
 [%steps]
 * Essayez la commande `cat README.md | cowsay` dans GitPod
-** Essayez de mettre à jour le workflow pour faire la même chose dans GitHub Actions
+** Modifiez l'étape "`cat README.md`" du workflow pour faire la même chose dans GitHub Actions
 ** SPOILER: ❌ (la commande `cowsay` n'est pas disponible dans le runner GitHub Actions)
 
-== Exercice : Environnement d'exécution
+== Environnement d'exécution
+
+* *Problème* : On souhaite utiliser les mêmes outils dans notre workflow ainsi que dans nos environnement de développement
+
+* Deux types de solutions existent, chacune avec ses avantages / inconvénients :
+** Personnaliser l'environnement dans votre workflow: (⚠️ sensible aux mises à jour)
+** Spécifier un environnement préfabriqué pour le workflow (✅ portable)
 
-* *But* : On souhaite utiliser une commande spécifique durant notre job
+== Exercice : Personnalisation dans le workflow
 
-* Deux types de solutions existent, chacune avec ses inconvénients :
-** Installer les outils manquants en préambule de chaque job (❌ lent ✅ facile )
-** Utiliser Docker pour fabriquer une action Github (❌ complexe ✅ portable)
+* *But* : exécuter la commande `cat README.md | cowsay` dans le workflow comme dans GitPod
 
-* C'est à vous :
-** Cherchez comment installer `cowman` dans Ubuntu 18.04
-** Appliquer cette solution dans votre job afin de le "réparer" et de voir la vache dans GitHub Actions.
+* C'est à vous de mettre à jour le workflow pour personnaliser l'environnement :
+** 💡 Cherchez comment installer `cowsay` dans le runner GitHub (`runs-on`...)
 
-== Solution : Environnement d'exécution
+== Solution : Personnalisation dans le workflow
 
 [source,yaml]
 ----
-include::../code-samples/gh-actions/say-hello-full.yml[tags="common,checkout,cowsay"]
+include::../code-samples/gh-actions/say-hello-full.yml[tags="common,checkout,apt-cowsay,run-cowsay"]
 ----
+
+== Exercice : Environnement préfabriqué
+
+* *But* : exécuter la commande `cat README.md | cowsay` dans le workflow comme dans GitPod
+** En utilisant le même environnement que GitPod (même version de cowsay, java, etc.)
+
+* C'est à vous de mettre à jour le workflow pour exécuter les étapes dans la même image Docker que GitPod :
+** 💡 link:https://github.com/cicd-lectures/gitpod/blob/2095c21e427aa4c1253958c3a9ba80394300dcd4/.gitpod.yml#L6[Image utilisée dans GitPod,window="_blank"]
+** 💡 link:https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#jobsjob_idcontainer[Utilisation d'un container comme runner GitHub Actions,window="_blank"]
+** 💡 link:https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#docker-container-filesystem[Contraintes d'exécution de container dans GitHub Actions (`--user=root`),window="_blank"]
+
+== Solution : Environnement préfabriqué
```

Comment:
Something that bothers me: we never introduce the idea of "action" as well, only mentions runs. That would be a third way to personalize the ci environment. WDYT?

Reply:
That's litterrally what I was drafting when I saw your comment \o/ 
We're mind-connected  ;)

I'm drafting a 3rd exercice (before the whole maven package) to illustrate a 3rd solution: "use your CI engine features" to ensure the best of both worlds (from the 2 other solutions).
In this case: GitHub Actions -> specific java action.

It just made me realize that the timing concept we discussed (e.g. get a feedback as soon as posible) should also be introduced in the CI chapter, before this 3rd exercise (that could be between #2 - using container image- and #3 by making them realize the impact on the build), with the following challenge statement:
- Start by building something that works
- Then optimize, with the focus of decreasing the time between a change and the feedback (improving the CI's feedback loop)

WDYT?

Reply:
I've added https://github.com/cicd-lectures/slides/pull/54/commits/559ef3de5c072ff49293c230a5a9a3caed92d3af with the draft content (I'm working on the exercise solution + tips, but it should give you an idea). WDYT?